cli/src/action/certs.rs

```diff
@@ -0,0 +1,389 @@
+
+    // return private key and ca_cert
+    Ok((privkey, cert))
+}
+
+// Make a certificate and private key signed by the given CA cert and private key
+// Cert could act like both server or client
+fn make_ca_signed_cert(
+    ca_cert: &X509Ref,
+    ca_privkey: &PKeyRef<Private>,
+    common_name: &str,
+) -> Result<(PKey<Private>, X509), CliError> {
+    // generate private key
+    let rsa = Rsa::generate(2048)?;
+    let privkey = PKey::from_rsa(rsa)?;
+
+    // build x509_name
+    let mut x509_name = X509NameBuilder::new()?;
+    x509_name.append_entry_by_text("CN", &common_name)?;
+    let x509_name = x509_name.build();
+
+    // build x509 cert
+    let mut cert_builder = X509::builder()?;
+    cert_builder.set_version(2)?;
+    let serial_number = {
+        let mut serial = BigNum::new()?;
+        serial.rand(159, MsbOption::MAYBE_ZERO, false)?;
+        serial.to_asn1_integer()?
+    };
+    cert_builder.set_serial_number(&serial_number)?;
+    cert_builder.set_subject_name(&x509_name)?;
+    cert_builder.set_issuer_name(ca_cert.subject_name())?;
+    cert_builder.set_pubkey(&privkey)?;
+    let not_before = Asn1Time::days_from_now(0)?;
+    cert_builder.set_not_before(&not_before)?;
+    let not_after = Asn1Time::days_from_now(365)?;
+    cert_builder.set_not_after(&not_after)?;
+
+    // allow keys to be used for both server and client authorization
+    cert_builder.append_extension(
+        ExtendedKeyUsage::new()
+            .server_auth()
+            .client_auth()
+            .build()?,
+    )?;
+
+    // sign the cert by the ca
+    cert_builder.sign(&ca_privkey, MessageDigest::sha256())?;
+    let cert = cert_builder.build();
+
+    // return private key and cert
+    Ok((privkey, cert))
+}
+
+/// write the a file to a temp file name and then rename to final file key_name
+/// this will guarantee that the final file will ony every contain valid data
+fn write_file(
+    path_buf: PathBuf,
+    file_name: &str,
+    bytes: &[u8],
+    change_permissions: bool,
+) -> Result<(), CliError> {
+    let temp_path_buf = path_buf.join(format!(".{}.new", file_name));
+    let temp_path = {
+        if let Some(path) = temp_path_buf.to_str() {
+            path.to_string()
+        } else {
+            return Err(CliError::ActionError(
+                "Path is not valid unicode".to_string(),
+            ));
+        }
+    };
+
+    let final_path_buf = path_buf.join(file_name);
+    let final_path = {
+        if let Some(path) = final_path_buf.to_str() {
+            path.to_string()
+        } else {
+            return Err(CliError::ActionError(
+                "Path is not valid unicode".to_string(),
+            ));
+        }
+    };
+
+    let mut file = OpenOptions::new()
+        .write(true)
+        .create(true)
+        .truncate(true)
+        .mode(0o640)
+        .open(temp_path.clone())?;
+    file.write_all(bytes)?;
+    fs::rename(temp_path, final_path)?;
+
+    if change_permissions {
```

Comment:
This should happen prior to the atomic rename above.

Reply:
fixed